arch/risc-v/src/mpfs/mpfs_can.c

```diff
@@ -0,0 +1,2923 @@
+ * Name: mpfs_ifdown
+ *
+ * Description:
+ *  NuttX Callback: Stop the CAN interface
+ *
+ * Input Parameters:
+ *  dev  - Reference to the NuttX driver state structure
+ *
+ * Returned Value:
+ *  Zero (CAN_OK) on success; a negated errno value on failure
+ *
+ * Assumptions:
+ *  None
+ *
+ ****************************************************************************/
+
+static int mpfs_ifdown(struct net_driver_s *dev)
+{
+  mpfs_can_instance_t *priv =
+    (mpfs_can_instance_t *)dev->d_private;
+
+  mpfs_can_stop(priv);
+
+  priv->bifup = false;
+  return CAN_OK;
+}
+
+/****************************************************************************
+ * Name: mpfs_ioctl
+ *
+ * Description:
+ *  PHY ioctl command handler
+ *
+ * Input Parameters:
+ *  dev  - Reference to the NuttX driver state structure
+ *  cmd  - ioctl command
+ *  arg  - Argument accompanying the command
+ *
+ * Returned Value:
+ *  Zero (CAN_OK) on success; a negated errno value on failure
+ *
+ * Assumptions:
+ *  None
+ *
+ ****************************************************************************/
+
+#ifdef CONFIG_NETDEV_IOCTL
+static int mpfs_ioctl(struct net_driver_s *dev, int cmd, unsigned long arg)
+{
+#ifdef CONFIG_DEBUG_CAN_INFO
+  caninfo("IOCTL received | cmd: %d arg: %ld\n", cmd, arg);
+#endif
+
+#if defined(CONFIG_NETDEV_CAN_BITRATE_IOCTL) || \
+defined(CONFIG_NETDEV_CAN_FILTER_IOCTL)
+  mpfs_can_instance_t *priv =
+    (mpfs_can_instance_t *)dev->d_private;
+#endif
+  int ret;
+
+  switch (cmd)
+    {
+#ifdef CONFIG_NETDEV_CAN_BITRATE_IOCTL
+    case SIOCGCANBITRATE:
+
+      /* Get bitrate from the CAN controller */
+
+      {
+        struct can_ioctl_data_s *req =
+          (struct can_ioctl_data_s *)((uintptr_t)arg);
+        req->arbi_bitrate = priv->bitrate_value;
+        req->arbi_samplep = mpfs_can_get_sample_point(priv);
+        ret = CAN_OK;
+      }
+      break;
+
+    case SIOCSCANBITRATE:
+
+      /* Set bitrate of the CAN controller */
+
+      {
+        struct can_ioctl_data_s *req =
+          (struct can_ioctl_data_s *)((uintptr_t)arg);
+
+        /* Stop CAN controller */
+
+        mpfs_can_stop(priv);
+
+        if (CAN_OK != mpfs_can_set_bitrate(priv, req->arbi_bitrate))
+          {
+            canerr("CAN controller bitrate set failed");
+            ret = -1;
+            break;
+          }
+
+        /* Start CAN controller again */
+
+        mpfs_can_start(priv);
+
+        ret = CAN_OK;
+      }
+      break;
+#endif /* CONFIG_NETDEV_CAN_BITRATE_IOCTL */
+
+#ifdef CONFIG_NETDEV_CAN_FILTER_IOCTL
+    case SIOCACANSTDFILTER:
+    case SIOCACANEXTFILTER:
+
+      {
+        struct can_ioctl_filter_s *req =
+          (struct can_ioctl_filter_s *)((uintptr_t)arg);
+
+        if (CAN_OK != mpfs_can_add_filter(priv, req->ftype,
+                                          req->fid1, req->fid2))
+          {
+            canerr("CAN filter add failed");
+            ret = -1;
```

Comment:
You should use a negative `errno` value here to be more descriptive. The upper-half `ioctl` will read that error code, set `errno = -ret` and then return -1. -1 should not be returned in the lower-half.